### PR TITLE
IconUrl deprecation message.

### PR DESCRIPTION
**Pull request by agr**
Addresses https://github.com/NuGet/NuGetGallery/issues/7478

Browser:
iconUrlDepr

Command line:
iconUrlDeprCL

### CODE REVIEW

File: src/NuGetGallery/Services/IconUrlDeprecationValidationMessage.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery
+{
+    /// <summary>
+    /// IconUrl element deprecation message.
```
Comment (scottbommarito) [inline src/NuGetGallery/Services/IconUrlDeprecationValidationMessage.cs]: nit: Not a very useful comment, as it just restates the name of the class. I would prefer either getting rid of this comment or saying more about IconUrl and its deprecation.

File: tests/NuGetGallery.Facts/Services/PackageUploadServiceFacts.cs [hunk 2]
```diff
@@ -1141,6 +1141,53 @@ public async Task AcceptsPackagesWithEmbeddedIconForFlightedUsers()
...
+                var result = await _target.ValidateBeforeGeneratePackageAsync(
+                    _nuGetPackage.Object,
+                    GetPackageMetadata(_nuGetPackage),
+                    _currentUser);
+
+                Assert.Equal(PackageValidationResultType.Accepted, result.Type);
+                Assert.Null(result.Message);
+                var warning = Assert.Single(result.Warnings);
+                Assert.IsType<IconUrlDeprecationValidationMessage>(warning);
+                Assert.StartsWith("<iconUrl> element will be deprecated, please consider switching to specifying the icon in the package.", warning.PlainTextMessage);
```
Comment (zhhyu) [inline tests/NuGetGallery.Facts/Services/PackageUploadServiceFacts.cs]: It looks easy to break this test if we use the hard coded string. Maybe directly using the "Strings.UploadPackage_IconUrlDeprecated" is better for future maintenances. 
I find some other examples in this facts.

Comment (agr) [reply tests/NuGetGallery.Facts/Services/PackageUploadServiceFacts.cs]: That's the point of this check. To break if text changes. Using `Strings.UploadPackage_IconUrlDeprecated` directly will hide such changes.

**Comment by karann-msft** [comment]
Can I see a mock with all possible warnings? I am afraid we may have one link for every warning and then another read more at the end which might be few too many warnings. We can choose to not fix this right now in which case the warning message looks fine. I'd just make sure it aligns with what I'd see on nuget pack.

**Comment by agr** [comment]
@karann-msft , this is the second *warning* message with the link (the other one is about license URL deprecation). All other warnings don't have links. And all other messages with link are errors (which we don't display at the same time as warnings).

File: src/NuGetGallery/Strings.resx [hunk 3]
```diff
@@ -1150,4 +1150,7 @@ The {1} Team</value>
   <data name="SiteAdminNotLoggedInWithRequiredTenant" xml:space="preserve">
     <value>The site admins are required to sign in with the '{0}' tenant only.</value>
   </data>
+  <data name="UploadPackage_IconUrlDeprecated" xml:space="preserve">
+    <value>&lt;iconUrl&gt; element will be deprecated, please consider switching to specifying the icon in the package.</value>
```
Comment (loic-sharma) [inline src/NuGetGallery/Strings.resx]: Considering that we're warning on icon URLs already, isn't it already deprecated? What do you think of this messaging?

```suggestion
    <value>The &lt;iconUrl&gt; element is deprecated, please use the icon element instead.</value>
```

This is message doesn't mirror the client's message, but I would prefer a clearer message if possible.

**Comment by karann-msft** [comment]
Let's update both messages to say "is deprecated". Rest looks good for now.

File: src/NuGetGallery/Strings.resx (hunk 3 above)
Comment (agr) [reply src/NuGetGallery/Strings.resx]: The message that client currently produces is:
> WARNING: NU5048: The 'PackageIconUrl'/'iconUrl' element is deprecated. Consider using the 'PackageIcon'/'icon' element instead. Learn more at https://aka.ms/deprecateIconUrl

And for `licenseUrl`:
> WARNING: NU5125: The 'licenseUrl' element will be deprecated. Consider using the 'license' element instead.

**Comment by agr** [comment]
Like that?
image

File: tests/NuGetGallery.Facts/Services/PackageUploadServiceFacts.cs (hunk 2 above)
Comment (dominoFire) [reply tests/NuGetGallery.Facts/Services/PackageUploadServiceFacts.cs]: > Like that?
> image
> 
> Saying that element is deprecated when we didn't release a "stable" version of the client supporting it sounds a bit weird though.

@agr Do you need to update your tests?

**Comment by loic-sharma** [comment]
> Saying that element is deprecated when we didn't release a "stable" version of the client supporting it sounds a bit weird though.

Does the client in .NET Core 3 warn? If so, that ships on the 23rd

**Pull request merged**